src/GraphQL/Instrumentation/GraphQLTelemetryProvider.cs

```diff
@@ -0,0 +1,149 @@
+/// </para>
+/// </summary>
+public class GraphQLTelemetryProvider : IConfigureExecution
+{
+    private readonly GraphQLTelemetryOptions _telemetryOptions;
+    private const string ACTIVITY_OPERATION_NAME = "graphql";
+
+    /// <summary>
+    /// Returns an <see cref="System.Diagnostics.ActivitySource"/> instance to be used for GraphQL.NET telemetry.
+    /// </summary>
+    protected static ActivitySource ActivitySource { get; }
+
+    /// <summary>
+    /// Gets the source name used for telemetry.
+    /// </summary>
+    public static string SourceName => "GraphQL";
+
+    static GraphQLTelemetryProvider()
+    {
+        ActivitySource = new ActivitySource(SourceName, typeof(GraphQLTelemetryProvider).Assembly.GetNuGetVersion());
+    }
+
+    /// <summary>
+    /// Initializes a new instance with the specified options.
+    /// </summary>
+    public GraphQLTelemetryProvider(GraphQLTelemetryOptions options)
+    {
+        _telemetryOptions = options;
+    }
+
+    /// <inheritdoc/>
+    public virtual float SortOrder => GraphQLBuilderExtensions.SORT_ORDER_CONFIGURATION;
+
+    /// <inheritdoc/>
+    public virtual async Task<ExecutionResult> ExecuteAsync(ExecutionOptions options, ExecutionDelegate next)
+    {
+        // start the Activity, in fact Activity.Stop() will be called from within Activity.Dispose() at the end of using block 
+        using var activity = await StartActivityAsync(options).ConfigureAwait(false);
+
+        // do not record any telemetry if there are no listeners or it decided not to sample the current request
+        if (activity == null)
+            return await next(options).ConfigureAwait(false);
+
+        // record the requested operation name and optionally the GraphQL document
+        await SetInitialTagsAsync(activity, options).ConfigureAwait(false);
+
+        // record the operation type and the operation name (which may be specified within the
+        // document even if not specified in the request)
+        options.Listeners.Add(new TelemetryListener(this, activity, options));
+
+        // execute the request
+        var result = await next(options).ConfigureAwait(false);
+
+        // record the status
+        await SetResultTagsAsync(activity, options, result).ConfigureAwait(false);
+
+        // return the result
+        return result;
+    }
+
+    /// <summary>
+    /// Creates an <see cref="Activity"/> for the specified <see cref="ExecutionOptions"/> and starts it.
+    /// </summary>
+    protected virtual ValueTask<Activity?> StartActivityAsync(ExecutionOptions options)
+        => new(ActivitySource.StartActivity(ACTIVITY_OPERATION_NAME));
+
+    /// <summary>
+    /// Sets the <see cref="Activity"/> tags based on the specified <see cref="ExecutionOptions"/>.
+    /// The default implementation sets the <c>graphql.operation.name</c> and <c>graphql.document</c> tags.
+    /// </summary>
+    protected virtual Task SetInitialTagsAsync(Activity activity, ExecutionOptions options)
```

Comment:
In general I would like to have the same design as for `AspNetCoreInstrumentationOptions` - switching to options with delegates and calling these delegates from `GraphQLTelemetryProvider` rather then force developer to write subclass to override adding additional info to activity. Options in this regard is more flexible. They may be initialized with default actions.